### PR TITLE
SNOW-786887: implement df.to_pandas

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/mock/connection.py [hunk 1]
```diff
@@ -539,19 +563,33 @@ def get_result_query_id(self, plan: SnowflakePlan, **kwargs) -> str:
...
+def _fix_pandas_df_integer(table_res: TableEmulator) -> "pandas.DataFrame":
+    pd_df = pd.DataFrame()
+    for col_name in table_res.columns:
+        col_sf_type = table_res.sf_types[col_name]
+        pd_df_col_name = unquote_if_quoted(col_name)
         if (
-            FIELD_ID_TO_NAME.get(column_metadata.type_code) == "FIXED"
-            and column_metadata.precision is not None
-            and column_metadata.scale == 0
-            and not str(pandas_dtype).startswith("int")
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/connection.py:552]: Why are we discarding `and not str(pandas_dtype).startswith("int")`?

File: src/snowflake/snowpark/mock/connection.py [hunk 2]
```diff
@@ -539,19 +563,33 @@ def get_result_query_id(self, plan: SnowflakePlan, **kwargs) -> str:
...
+            elif col_sf_type.datatype.precision <= 4:
+                pd_df[pd_df_col_name] = table_res[col_name].astype("int16")
+                # pd_df[pd_df_col_name].dtype = "int16"
+            elif col_sf_type.datatype.precision <= 8:
+                pd_df[pd_df_col_name] = table_res[col_name].astype("int32")
+                # pd_df[pd_df_col_name].dtype = "int32"
+            else:
+                pd_df[pd_df_col_name] = table_res[col_name].astype("int64")
+                # pd_df[pd_df_col_name].dtype = "int64"
+            # pd_df[pd_df_col_name].sf_type = table_res[col_name].sf_type
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/connection.py:587]: Shall we remove this line?

File: src/snowflake/snowpark/mock/connection.py [hunk 3]
```diff
@@ -539,19 +563,33 @@ def get_result_query_id(self, plan: SnowflakePlan, **kwargs) -> str:
...
         if (
-            FIELD_ID_TO_NAME.get(column_metadata.type_code) == "FIXED"
-            and column_metadata.precision is not None
-            and column_metadata.scale == 0
-            and not str(pandas_dtype).startswith("int")
+            isinstance(col_sf_type.datatype, DecimalType)
+            and col_sf_type.datatype.precision is not None
+            and col_sf_type.datatype.scale == 0
         ):
-            pd_df[pandas_col_name] = pandas.to_numeric(
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/connection.py:554]: Question: why can't we use the same code here?

File: src/snowflake/snowpark/mock/functions.py [hunk 4]
```diff
@@ -525,9 +525,11 @@ def mock_to_timestamp(
             if data is None:
                 res.append(None)
                 continue
-            if auto_detect and data.isnumeric():
+            if auto_detect and (
+                isinstance(data, int) or (isinstance(data, str) and data.isnumeric())
+            ):
                 res.append(
-                    datetime.datetime.utcfromtimestamp(process_numeric_time(data))
+                    datetime.datetime.utcfromtimestamp(process_numeric_time(int(data)))
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/functions.py:532]: We don't have to cast to `int` here, `process_numeric_time` is expecting a `str` and will cast that to `int`.

File: tests/mock/test_to_pandas.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+from snowflake.snowpark import Session
+from snowflake.snowpark.mock.connection import MockServerConnection
+
+session = Session(MockServerConnection())
+
+
+@pytest.mark.localtest
+def test_df_to_pandas_df():
+    df = session.create_dataframe(
```
Comment (sfc-gh-stan) [inline tests/mock/test_to_pandas.py:19]: Can we cover more types, i.e. https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/latest/types except for Geography and Geometry?

File: src/snowflake/snowpark/mock/connection.py [hunk 6]
```diff
@@ -405,8 +410,27 @@ def execute(
...
+            for col_name in res.columns:
+                pandas_df[unquote_if_quoted(col_name)] = res[col_name].tolist()
+            rows = _fix_pandas_df_integer(res)
+
+            # the following implementation is just to make DataFrame.to_pandas_batches API workable
+            # in snowflake, large data result are split into multiple data chunks
+            # and sent back to the client, thus it makes sense to have the generator
+            # however, local testing is designed for local testing
+            # we do not mock the splitting into data chunks behavior
+            if to_iter:
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/connection.py:424]: Why can't we reuse the if block on line 431?

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/connection.py:424]: document the generator vs iter behavior upon PD frame

File: src/snowflake/snowpark/mock/connection.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/connection.py:552]: this is to align the behavior with live connection to optimize dtype integer.
previous assumption was wrong, `int64` type also needs to be optimized so I remove the condition

File: src/snowflake/snowpark/mock/connection.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/connection.py:554]: pandas.to_numeric doesn't performa the type cast

File: src/snowflake/snowpark/mock/connection.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/connection.py]: updated

File: src/snowflake/snowpark/mock/functions.py (hunk 4 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/functions.py]: updated

File: tests/mock/test_to_pandas.py (hunk 5 above)
Comment (sfc-gh-aling) [reply tests/mock/test_to_pandas.py:40]: updated

File: src/snowflake/snowpark/mock/connection.py [hunk 7]
```diff
@@ -405,8 +410,29 @@ def execute(
...
+                # we return a generator instead of iter(rows) because in the live connection
+                # rows is pd.DataFrame object, iter(rows) will return rows inside the pd.DataFrame which diverges
+                # from the live connection
+                def gen():
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/connection.py:428]: I think we can replace this with `rows = [rows]`, then later code on line 433 can be reused.

**Review by sfc-gh-stan — APPROVED** [review]
LGTM with one comment.

File: src/snowflake/snowpark/mock/connection.py (hunk 7 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/connection.py:428]: good idea! code updated

**Pull request merged**